src/Features/Core/Portable/Completion/Providers/AbstractEnumCompletionProvider.cs

```diff
@@ -0,0 +1,142 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Recommendations;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions.ContextQuery;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers
+{
+    internal abstract class AbstractEnumCompletionProvider : AbstractSymbolCompletionProvider
+    {
+        private static readonly CompletionItemRules s_rules = CompletionItemRules.Default.WithMatchPriority(MatchPriority.Preselect);
+
+        protected abstract (string displayText, string suffix, string insertionText) GetDefaultDisplayAndSuffixAndInsertionText(ISymbol symbol, SyntaxContext context);
+
+        protected override Task<ImmutableArray<ISymbol>> GetPreselectedSymbolsAsync(SyntaxContext context, int position, OptionSet options, CancellationToken cancellationToken)
+        {
+            var syntaxFacts = context.GetLanguageService<ISyntaxFactsService>();
+            if (syntaxFacts.IsInNonUserCode(context.SyntaxTree, context.Position, cancellationToken))
+                return SpecializedTasks.EmptyImmutableArray<ISymbol>();
+
+            // This providers provides fully qualified names, eg "DayOfWeek.Monday"
+            // Don't run after dot because SymbolCompletionProvider will provide
+            // members in situations like Dim x = DayOfWeek.$$
+            if (context.TargetToken.RawKind == syntaxFacts.SyntaxKinds.DotToken)
+                return SpecializedTasks.EmptyImmutableArray<ISymbol>();
+
+            var typeInferenceService = context.GetLanguageService<ITypeInferenceService>();
+            var enumType = typeInferenceService.InferType(context.SemanticModel, position, objectAsDefault: true, cancellationToken: cancellationToken);
+
+            if (enumType is not { TypeKind: TypeKind.Enum })
```

Comment:
```suggestion
            if (enumType?.TypeKind != TypeKind.Enum)
```

Reply:
i'm not opposed to patterns.  but when there is already a clear, idiomatic, shorter non-pattern form, i would prefer that.